### ISSUE TITLE
Repeated non-finite grad_norm and reproduction

### ISSUE THREAD

**Issue by luc-casia** [LeapLabTHU/ViTTT#11]
Hi, thank you for releasing ViTTT. I encountered repeated non-finite gradient norms when training ViTTT-T with AMP.

### Environment
- Python 3.9
- PyTorch 1.11.0
- torchvision 0.12.0
- timm 0.4.12
- 4 GPUs
- batch size per GPU: 1024
- total batch size: 4096
- scaled base learning rate: 0.004
- AMP enabled

After some epochs, I repeatedly observe logs such as:
Train: [88/300][100/312] grad_norm 0.8387 (0.6661)
Train: [88/300][200/312] grad_norm 0.6910 (nan)
Train: [88/300][300/312] grad_norm 0.7382 (nan)

This suggests that one or more unlogged batches may have produced a non-finite gradient norm. The loss remains finite and the validation accuracy continues to improve. However, it did not achieve the expected results as anticipated in the final paper.

Could you please clarify:
Did you observe this grad_norm NaN issue when training the models reported in the paper?
Were the reported models trained with AMP enabled?
Is this behavior expected, and can it affect the final ImageNet accuracy?

Thank you!

**Comment by tian-qing001** [comment]
Hi @luc-casia, thanks for the detailed information!

The models reported in the paper were trained with AMP enabled. In this setting, seeing `grad_norm` become `NaN` is expected. When training models on ImageNet with AMP enabled, `grad_norm` becoming `NaN` is common for most models. Typically, enabling AMP does not significantly affect the final accuracy. However, if training time is not a major concern, you could try disabling AMP and see whether this improves the results.

To help with reproduction, we have also provided our original training logs [here](https://cloud.tsinghua.edu.cn/f/7daa0c48db8242ce83c5/). One thing to note is that the released code has been cleaned up and some variable names have been renamed, so there may be some differences between the original logs and the current version of the code.

**Comment by luc-casia** [comment]
> Hi [@luc-casia](https://github.com/luc-casia), thanks for the detailed information!
> 
> The models reported in the paper were trained with AMP enabled. In this setting, seeing `grad_norm` become `NaN` is expected. When training models on ImageNet with AMP enabled, `grad_norm` becoming `NaN` is common for most models. Typically, enabling AMP does not significantly affect the final accuracy. However, if training time is not a major concern, you could try disabling AMP and see whether this improves the results.
> 
> To help with reproduction, we have also provided our original training logs [here](https://cloud.tsinghua.edu.cn/f/7daa0c48db8242ce83c5/). One thing to note is that the released code has been cleaned up and some variable names have been renamed, so there may be some differences between the original logs and the current version of the code.

Thank you very much for sharing the full training log! It has been extremely helpful for checking the training configuration and learning‑rate schedule!

I reproduced the experiment using the public code with Python 3.9, PyTorch 1.11.0, torchvision 0.12.0, and timm 0.4.12 on 8 A100 GPUs. I used a per‑GPU batch size of 512, giving a global batch size of 4096. I also kept the base learning rate at 5e‑4 before the automatic linear scaling in main.py, so that the effective learning rate printed in the log was 0.004, matching the shared log.

The training completed normally, but I still observe a non‑negligible accuracy gap compared with the result in your log. Given that I have matched all the hyperparameters I could find, I would like to ask one major question that I hope you can help with: Could you please confirm whether the logged experiment used the exact public code revision, and whether there were any additional modifications, uncommitted patches, launch arguments, or non‑default training settings used for that run?

I would be extremely grateful for any clarification. Thank you again for your time and for providing the detailed log; I greatly appreciate any guidance you can offer.

**Comment by tian-qing001** [comment]
@luc-casia Could you please share your complete training log for this run? I will check it against my reproduction.

**Comment by luc-casia** [comment]
Thanks for your follow-up and for offering to check the log!

Regarding the training log you requested: unfortunately, I am unable to export the log from the 8×A100 run (total batch size 4096) due to internal access restrictions on that machine.

What I do have is the complete training log from a 4×A800 run with a total batch size of 1024 (per‑GPU 256). I should clarify one thing: the accuracy from this 4×A800 run is actually about 0.2% higher than what I got from the 8×A100 run with BS=4096. So if you use this log for comparison, please keep this difference in mind.

I have uploaded the 4×A800 log file below – you can download it directly.
[log_rank0.txt](https://github.com/user-attachments/files/31859281/log_rank0.txt)

If you would prefer a log that better matches your setup (i.e., total batch size 4096), I am also happy to rerun the experiment on 8×A800 with BS=4096 and share that log with you instead. Just let me know and I will do it.

Thank you again for your time and help!

**Comment by tian-qing001** [comment]
Hi @luc-casia, thanks a lot for your reproduction and detailed feedback. We really appreciate you digging into this issue carefully.

We have reproduced the training with the current public codebase, and we also observed the accuracy gap compared with the results reported in the paper, which is consistent with your finding.

**Comment by tian-qing001** [comment]
@luc-casia 

We further investigated this gap and found that **it is caused by a difference in weight initialization, not a bug in the model architecture or training logic**.

During the code cleanup for public release, we removed several initialization calls for parameters that are not used in the final version of the TTT module. Although these parameters themselves do not participate in the final forward computation, they consume the global PyTorch random number generator (RNG) state during initialization. Removing them shifts the RNG state, which in turn changes the initialization of all subsequent weights.

**Comment by tian-qing001** [comment]
@luc-casia 

### How to reproduce the exact paper results
If you want to perfectly align the initialization state with the original experiments in the paper, you can insert the following lines [here](https://github.com/LeapLabTHU/ViTTT/blob/e3477587d099e6b9e83e9e7c80b1b999e0989a20/vittt/models/ttt_block.py#L38).

```python
# Align RNG consumption with original training code to align initialization
self.ape = nn.Parameter(torch.zeros(1, 196, dim))
trunc_normal_(self.ape, std=.02)
del self.ape
self.qkv = nn.Linear(dim, dim * 3, bias=qkv_bias)
self.proj = nn.Linear(dim, dim)
```

These lines have no effect on the final model structure or computation. The parameters are either overwritten by the real module weights or deleted immediately. They only serve to align the number of RNG draws, so that all subsequent weight initializations match exactly the ones used in the paper.

After aligning the RNG state this way, the training results become fully consistent with our original runs. We offer the full training log of this successful reproduction for your reference:
[log.txt](https://cloud.tsinghua.edu.cn/f/a13bd7f90e5b4dddab42/)

We were also surprised that such a difference in initialization could lead to a noticeable accuracy gap. We suspect this effect is limited to ViTTT-T.

**Comment by tian-qing001** [comment]
@luc-casia 

To summarize:
- The currently released code is correct in model structure, training algorithm and computation flow, and faithfully implements the method described in the paper.
- The small discrepancy between reproduced results and reported numbers comes from the RNG state shift introduced during code cleanup.
- Due to limited compute resources, we will not be conducting further experiments specifically on this initialization detail.

Thanks again for your careful verification and valuable discussion.
